File: creational/factory-method/factory-method.ts
enum AnimalBreeds {
    siamese,
    maineCoon,
    siberianHusky
}

abstract class AnimalData {
    protected abstract _breed: string;
    protected abstract _origin: string;

    public composeBreedData(): string {
        return `Breed: ${this._breed}`;
    }

    public composeOriginData(): string {
        return `From: ${this._origin}`;
    }
}

class SiameseData extends AnimalData {
    protected _breed: string = 'Siamese';
    protected _origin: string = 'Thailand';
    
}

class MaineCoonData extends AnimalData {
    protected _breed: string = 'Maine Coon';
    protected _origin: string = 'Maine, United States of America';
}

class SiberianHuskyData extends AnimalData {
    protected _breed: string = 'Siberian Husky';
    protected _origin: string = 'Siberia';
}

abstract class AnimalDatabase {
    protected abstract _makeQuery(breed: AnimalBreeds): AnimalData;

    public searchByBreed(breed: AnimalBreeds): string {
        let animal: AnimalData = this._makeQuery(breed);

        return `${animal.composeBreedData()}\n${animal.composeOriginData()}`;
    }
}

class CatDatabase extends AnimalDatabase {
    protected _makeQuery(breed: AnimalBreeds): AnimalData {
        if (breed === AnimalBreeds.siamese) {
            return new SiameseData();
        } else if (breed === AnimalBreeds.maineCoon) {
            return new MaineCoonData();
        } else {
            return null;
        }
    }
}

class DogDatabase extends AnimalDatabase {
    protected _makeQuery(breed: AnimalBreeds): AnimalData {
        if (breed === AnimalBreeds.siberianHusky) {
            return new SiberianHuskyData();
        } else {
            return null;
        }
    }
}

let catDatabase: AnimalDatabase = new CatDatabase();
let dogDatabase: AnimalDatabase = new DogDatabase();

console.log(catDatabase.searchByBreed(AnimalBreeds.siamese));
console.log(catDatabase.searchByBreed(AnimalBreeds.maineCoon));
console.log(dogDatabase.searchByBreed(AnimalBreeds.siberianHusky));

File: creational/factory-method/factory-method.cs
using System;

class FactoryMethodPattern
{
    enum AnimalBreeds
    {
        siamese,
        maineCoon,
        siberianHusky
    }

    abstract class AnimalData
    {
        protected abstract string breed { get; }
        protected abstract string origin { get; }

        public string ComposeBreedData()
        {
            return String.Format("Breed: {0}", breed);
        }

        public string ComposeOriginData()
        {
            return String.Format("From: {0}", origin);
        }
    }

    class SiameseData : AnimalData
    {
        protected override string breed { get; } = "Siamese";
        protected override string origin { get; } = "Thailand";
        
    }

    class MaineCoonData : AnimalData
    {
        protected override string breed { get; } = "Maine Coon";
        protected override string origin { get; } = "Maine, United States of America";
    }

    class SiberianHuskyData : AnimalData
    {
        protected override string breed { get; } = "Siberian Husky";
        protected override string origin { get; } = "Siberia";
    }

    abstract class AnimalDatabase
    {
        protected abstract AnimalData MakeQuery(AnimalBreeds breed);

        public string SearchByBreed(AnimalBreeds breed)
        {
            AnimalData animal = MakeQuery(breed);

            return String.Format("{0}\n{1}", animal.ComposeBreedData(), animal.ComposeOriginData());
        }
    }

    class CatDatabase : AnimalDatabase
    {
        protected override AnimalData MakeQuery(AnimalBreeds breed)
        {
            if (breed == AnimalBreeds.siamese)
            {
                return new SiameseData();
            }
            else if (breed == AnimalBreeds.maineCoon)
            {
                return new MaineCoonData();
            }
            else
            {
                return null;
            }
        }
    }

    class DogDatabase : AnimalDatabase
    {
        protected override AnimalData MakeQuery(AnimalBreeds breed)
        {
            if (breed == AnimalBreeds.siberianHusky)
            {
                return new SiberianHuskyData();
            }
            else
            {
               return null;
            }
        }
    }

    static void Main(string[] args)
    {
        AnimalDatabase catDatabase = new CatDatabase();
        AnimalDatabase dogDatabase = new DogDatabase();

        Console.WriteLine(catDatabase.SearchByBreed(AnimalBreeds.siamese));
        Console.WriteLine(catDatabase.SearchByBreed(AnimalBreeds.maineCoon));
        Console.WriteLine(dogDatabase.SearchByBreed(AnimalBreeds.siberianHusky));
    }
}
